### ISSUE TITLE
Docker-in-docker does not clear docker data on rebuild

### ISSUE THREAD

**Issue by worthy7** [microsoft/vscode-remote-release#9193]
https://github.com/microsoft/vscode-remote-release/issues/5679

Related to this issue, where Version 2 will give unique names to each project.

The issue however - is that when rebuilding a dev container from scratch, it will reuse an existing docker installation and so any containers will automatically run. This does not mean a reproducable environment for a devcontainer and I believe just defeats the point of dev-containers altogether.

Currently:
docker-in-docker works by running a DinD container for each project

Expected/wanted:
docker-in-docker works by running a DinD container for each BUILD of the dev container.

**Comment by bamurtaugh** [comment]
Thank you for filing!

@samruddhikhandale @chrmarti I'd be curious to get your thoughts on this one.

**Comment by worthy7** [comment]
At the moment my current work around is that before destroying the container I run 
`docker stop $(docker ps -a -q);docker remove $(docker ps -a -q);docker volume remove $(docker volume list -q)`

If I forget, it does weird things like when the devcontainer comes back, things that were mounted from devcontainer into internal containers before, become realized by the internal containers instead under root user and cannot be accessed by my vscode user anymore. That's just one symptom of this issue.

**Comment by worthy7** [comment]
A possible solution is that the `install.sh` removes the DinD volume used for the devcontainer's docker before recreating it.

**Comment by chrmarti** [comment]
We could consider making the volume optional, so the nested containers would be part of the dev container and get deleted on rebuild. Currently there is no way to include the volume depending on a feature's options, so we can't just add it there.

**Comment by worthy7** [comment]
> We could consider making the volume optional, so the nested containers would be part of the dev container and get deleted on rebuild. Currently there is no way to include the volume depending on a feature's options, so we can't just add it there.

Yes PLEASE this sounds ideal. An extra option "Use host docker volume?" and to be honest it should be off by default but maybe it will need to be `on` for compatibility purposes. 
Is it really "docker-in-docker" if it requires the docker outside of docker? 🤷‍♂️

**Comment by chrmarti** [comment]
I guess depending on how you intend to use docker-in-docker, you might expect the data (containers in this case) to be preserved across rebuilds of the dev container. I agree we would probably keep the volume by default for compatibility. (Note that this probably won't happen anytime soon as we don't have a way to make mount points conditional on a feature option.)

**Comment by worthy7** [comment]
I am interested to know why it's not just a folder inside the devcontainer in the first place..... currently I would have to just go and install docker myself inside a devcontainer to get what I want to achieve - and to be honest I think that absolutely nobody is expecting the internal docker containers to "live" though a rebuild of the devcontainer. 

I think v3 should have it removed
